examples/go/pagerduty-trigger/handler/handler.go

```diff
@@ -0,0 +1,202 @@
+	if err != nil {
+		return pdConfig{}, fmt.Errorf("read pdconfig.toml: %w", err)
+	}
+
+	if err := secret.Unmarshal(&pdc); err != nil {
+		return pdConfig{}, fmt.Errorf("unmarshal pdconfig.toml: %w", err)
+	}
+
+	if err := validatePdConf(pdc); err != nil {
+		return pdConfig{}, err
+	}
+
+	return pdc, nil
+}
+
+func validatePdConf(pdc pdConfig) error {
+	if pdc.PagerDuty.RoutingKey == "" {
+		return errors.New("PagerDuty routing key cannot be empty in pdconfig.toml")
+	}
+
+	if pdc.PagerDuty.EventAction == "" {
+		return errors.New("PagerDuty event action cannot be empty in pdconfig.toml")
+	}
+
+	return nil
+}
+
+// parseEvent saves info from the triggering event into a PagerDuty data object.
+func parseEvent(req []byte, pdc pdConfig) (pdPayload, error) {
+	var ce cloudEvent
+	var pd pdPayload
+
+	if err := json.Unmarshal(req, &ce); err != nil {
+		return pd, err
+	}
+
+	if err := validateEvent(ce); err != nil {
+		return pd, err
+	}
+
+	pd.RoutingKey = pdc.PagerDuty.RoutingKey
+	pd.EventAction = pdc.PagerDuty.EventAction
+	pd.Client = "VMware Event Broker Appliance"
+	pd.ClientURL = ce.Source
+	pd.Payload.Summary = ce.Data.FullFormattedMessage
+	pd.Payload.Timestamp = ce.Data.CreatedTime
+	pd.Payload.Source = ce.Source
+	pd.Payload.Severity = "info"
+	pd.Payload.Component = ce.Data.Vm.Name
+	pd.Payload.Group = ce.Data.Host.Name
+	pd.Payload.Class = ce.Subject
+	pd.Payload.CustomDetails.User = ce.Data.UserName
+	pd.Payload.CustomDetails.Datacenter = ce.Data.Datacenter
+	pd.Payload.CustomDetails.ComputeResource = ce.Data.ComputeResource
+	pd.Payload.CustomDetails.Host = ce.Data.Host
+	pd.Payload.CustomDetails.VM = ce.Data.Vm
+
+	return pd, nil
+}
+
+func validateEvent(event cloudEvent) error {
+	var msg string
+
+	switch true {
+	case event.Source == "":
+		msg = "does not contain Source"
+	case event.Subject == "":
+		msg = "does not contain Subject"
+	case event.Data.FullFormattedMessage == "":
+		msg = "does not contain Data.FullFormattedMessage"
+	case (event.Data.CreatedTime == time.Time{}):
+		msg = "does not contain Data.CreatedTime"
+	case event.Data.Vm.Name == "":
+		msg = "does not contain Data.Vm.Name"
+	case event.Data.Host.Name == "":
+		msg = "does not contain Data.Vm.Name"
+	}
+
+	if msg != "" {
+		return errors.New("invalid event: " + msg)
+	}
+
+	return nil
+}
+
+func pdSendRequest(ctx context.Context, path string, pdp pdPayload) ([]byte, error) {
+	tpCfg := &http.Transport{
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: os.Getenv("tls_insecure") == "true",
+		},
+	}
+
+	clt := &http.Client{
+		Transport: tpCfg,
+	}
+
+	reqBody, err := json.Marshal(pdp)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "POST", path, bytes.NewBuffer(reqBody))
+	if err != nil {
+		return []byte{}, fmt.Errorf("create http request: %w", err)
+	}
+
+	res, err := clt.Do(req)
+	if err != nil {
+		return []byte{}, fmt.Errorf("execute http request %w", err)
+	}
+	defer res.Body.Close()
```

Comment:
I would check for nil before closing on `res` and move this just after calling `clt.Do` to avoid null pointer

Reply:
This is a good thought exercise.

To clarify, is the following what you meant when you said "check for nil before closing on `res`"? 
```go
	res, err := clt.Do(req)
        if res == nil {
             return []byte{}, errors.New("failed POST request to PagerDuty API")
        }
	defer res.Body.Close()

        if err != nil {
		return []byte{}, fmt.Errorf("execute http request %w", err)
	}
```

In the way I wrote it originally, I checked for errors before `defer res.Body.Close()`. Do you think it's possible that `res.Body.Close()` might result in nil pointer if there was no error? I haven't done enough API requests to know.

Looking at the Go documentation, I found this section:
```
// The http Client and Transport guarantee that Body is always
// non-nil, even on responses without a body or responses with
// a zero-length body. 
```
Source: https://golang.org/pkg/net/http/#Response

The problem is, `res` can still be nil, right? Even if `Body` is "always non-nil". Do you think maybe `res` is also always non-nil necessarily, if Body is always non-nil?

Reply:
I found this (article)[http://devs.cloudimmunity.com/gotchas-and-common-mistakes-in-go-golang/#anameclose_http_resp_bodyaclosinghttpresponsebody] that I believe does what you suggest and gives good reasoning. I've changed the code to look like this:

```
res, err := clt.Post(path, "application/json", bytes.NewBuffer(reqBody))
	if res != nil {
		defer res.Body.Close()
	}

	if err != nil {
		return []byte{}, fmt.Errorf("execute http request %w", err)
	}
```

Thank you for helping me to learn this little Go trap and how to deal with it 😄